docs/user/customization.rst

```diff
@@ -96,3 +116,141 @@ perform some book keeping, like making sure that callbacks are handled
+   this:
+
+.. code:: python
+
+    net = MyNet(
+        module=MyModule,
+        module__num_units=100,
+
+        othermodule=MyOtherModule,
+        othermodule__num_units=200,
+    )
+    net.fit(X, y)
+
+A word about the distinction between modules and criteria made by skorch:
+Typically, criteria are also just subclasses of PyTorch
+:class:`~torch.nn.Module`. As such, skorch moves them to CUDA if that is the
+indicated device and will even pass parameters of criteria to the optimizers, if
+there are any. This can be useful when e.g. training GANs, where you might
+implement the discriminator as the criterion (and the generator as the module).
+
+A difference between module and criterion is that the output of modules are used
+for generating the predictions and are thus returned by
+:func:`~skorch.net.NeuralNet.predict` etc. In contrast, the output of the
+criterion is used for calculating the loss and should therefore be a scalar.
+
+skorch assumes that criteria may depend on the modules. Therefore, if a module
+is re-initialized, all criteria are also re-initialized, but not vice-versa. On
+top of that, the optimizer is re-initialized when either modules or criteria
+are changed.
+
+So after all this talk, what are the aforementioned guidelines to add your own
+modules, criteria, and optimizers? You have to follow these rules:
+
+1. Initialize them during their respective ``initialize_`` methods, e.g. modules
+   should be set inside :func:`~skorch.net.NeuralNet.initialize_module`.
+2. If they have learnable parameters, they should be instances of
+   :class:`~torch.nn.Module`. Optimizers should be instances of
+   :class:`~torch.optim.Optimizer`.
+3. Their names should end on an underscore. This is true for all attributes that
+   are created during ``initialize`` and distinguishes them from arguments
+   passed to ``__init__``. So a name for a custom module could be ``mymodule_``.
+4. Inside the initialization method, use :meth:`.get_params_for` (or,
+   if dealing with an optimizer, :meth:`.get_params_for_optimizer`) to
+   retrieve the arguments for the constructor of the instance.
+
+Here is an example of how this could look like in practice:
+
+.. code:: python
+
+    class MyNet(NeuralNet):
+        def initialize_module(self, *args, **kwargs):
+            super().initialize_module(*args, **kwargs)
+
+            # add an additional module called 'module2_'
+            params = self.get_params_for('module2')
+            self.module2_ = Module2(**params)
+            return self
+
+        def initialize_criterion(self, *args, **kwargs):
+            super().initialize_criterion(*args, **kwargs)
+
+            # add an additional criterion called 'other_criterion_'
+            params = self.get_params_for('other_criterion')
+            self.other_criterion_ = nn.BCELoss(**params)
+            return self
+
+        def initialize_optimizer(self, *args, **kwargs):
+            # first initialize the normal optimizer
```

Comment:
The example in the docstring for `get_learnable_params` says to override `get_learnable_params` and `initialize_optimizer` to do this. I think what we have here is simpler.

If a user adds their own `module2_`, they can still access them through `self.module2_.named_parameters`. What `get_learnable_params` provides is a nice wrapper to connect the optimizer and the module together, but it require a user to implement this connection.

Reply:
You are right, it is probably better to encourage users to directly use `module_.named_parameters()`. The `get_learnable_params` method could be relegated to basically being a convenience method to return all named parameters at once. We could therefore remove the `optimizer_name` argument, I'm on the fence what the best design would be here.